pkg/sources/manager.go

```diff
@@ -263,80 +299,136 @@ func (m *Manager) Run(ctx context.Context) {
+func (m *Manager) Sources(fn func(*SourceInfo), stats bool) {
 	done := make(chan struct{})
 	m.fns <- func(m *Manager) {
 		defer close(done)
+		si := new(SourceInfo)
 		for _, s := range m.sources {
-			fn(s.id, s.name, s.url, s.active, s.rate, s.slots, s.headers,
-				s.strictMode, s.insecure, s.signatureCheck, s.age, s.ignorePatterns,
-				s.clientCertPublic != nil,
-				s.clientCertPrivate != nil,
-				s.clientCertPassphrase != nil,
-			)
+			var st *Stats
+			if stats {
+				st = new(Stats)
+				s.addStats(st)
+			}
+			*si = SourceInfo{
+				ID:                      s.id,
+				Name:                    s.name,
+				URL:                     s.url,
+				Active:                  s.active,
+				Rate:                    s.rate,
+				Slots:                   s.slots,
+				Headers:                 s.headers,
+				StrictMode:              s.strictMode,
+				Insecure:                s.insecure,
+				SignatureCheck:          s.signatureCheck,
+				Age:                     s.age,
+				IgnorePatterns:          s.ignorePatterns,
+				HasClientCertPublic:     s.clientCertPublic != nil,
+				HasClientCertPrivate:    s.clientCertPrivate != nil,
+				HasClientCertPassphrase: s.clientCertPassphrase != nil,
+				Stats:                   st,
+			}
+			fn(si)
 		}
 	}
 	<-done
 }
 
 // Feeds passes the fields of the feeds of a given source to a given function.
-func (m *Manager) Feeds(sourceID int64, fn func(
-	id int64,
-	label string,
-	url *url.URL,
-	rolie bool,
-	lvl config.FeedLogLevel,
-)) error {
+func (m *Manager) Feeds(sourceID int64, fn func(*FeedInfo), stats bool) error {
 	errCh := make(chan error)
 	m.fns <- func(m *Manager) {
 		s := m.findSourceByID(sourceID)
 		if s == nil {
 			errCh <- NoSuchEntryError("no such source")
 			return
 		}
+		fi := new(FeedInfo)
 		for _, f := range s.feeds {
-			fn(f.id, f.label, f.url, f.rolie, config.FeedLogLevel(f.logLevel.Load()))
+			if f.invalid.Load() {
+				continue
+			}
+			var st *Stats
+			if stats {
+				st = new(Stats)
+				f.addStats(st)
+			}
+			*fi = FeedInfo{
+				ID:    f.id,
+				Label: f.label,
+				URL:   f.url,
+				Rolie: f.rolie,
+				Lvl:   config.FeedLogLevel(f.logLevel.Load()),
+				Stats: st,
+			}
+			fn(fi)
 		}
 		errCh <- nil
 	}
 	return <-errCh
 }
 
-// Feed passes the fields of feed to a given function.
-func (m *Manager) Feed(feedID int64, fn func(
-	label string,
-	url *url.URL,
-	rolie bool,
-	lvl config.FeedLogLevel,
-)) error {
-	errCh := make(chan error)
+// Feed returns the infos of a feed.
+func (m *Manager) Feed(feedID int64, stats bool) *FeedInfo {
+	fiCh := make(chan *FeedInfo)
 	m.fns <- func(m *Manager) {
 		f := m.findFeedByID(feedID)
-		if f == nil {
-			errCh <- NoSuchEntryError("no such feed")
+		if f == nil || f.invalid.Load() {
+			fiCh <- nil
 			return
 		}
-		fn(f.label, f.url, f.rolie, config.FeedLogLevel(f.logLevel.Load()))
-		errCh <- nil
+		var st *Stats
+		if stats {
+			st = new(Stats)
+			f.addStats(st)
+		}
```

Comment:
For feeds and sources, it can be useful to extend the function to only return the statistics. This can be beneficial, if we later implement polling of the statistics on the frontend.

Reply:
> For feeds and sources, it can be useful to extend the function to only return the statistics. This can be beneficial, if we later implement polling of the statistics on the frontend.

The other values are cheap enough to be polled.